cobalt/watchdog/watchdog.cc

```diff
@@ -579,23 +579,46 @@ std::string Watchdog::GetWatchdogViolations(
             violations_count_ -= violations_count;
             if (!static_cast<base::DictionaryValue*>(violations_map_.get())
                      ->empty()) {
-              WriteWatchdogViolations();
+              pending_write_ = true;
             } else {
               starboard::SbFileDeleteRecursive(GetWatchdogFilePath().c_str(),
                                                true);
             }
           }
         }
       }
-      if (!filtered_client_data.DictEmpty()) {
-        base::JSONWriter::Write(filtered_client_data, &watchdog_json_fetched);
+      if (!fetched_violations.DictEmpty()) {
+        base::JSONWriter::Write(fetched_violations, &fetched_violations_json);
+      }
+      if (clear) {
+        EvictOldWatchdogViolations();
       }
     }
-    SB_LOG(INFO) << "[Watchdog] Reading violations:\n" << watchdog_json_fetched;
+    SB_LOG(INFO) << "[Watchdog] Reading violations:\n"
+                 << fetched_violations_json;
   } else {
     SB_LOG(INFO) << "[Watchdog] No violations.";
   }
-  return watchdog_json_fetched;
+  return fetched_violations_json;
+}
+
+void Watchdog::EvictOldWatchdogViolations() {
+  int64_t current_timestamp = SbTimeToPosix(SbTimeGetNow()) / 1000;
+  int64_t cutoff_timestamp = current_timestamp - kWatchdogMaxViolationsAge;
+  for (const auto& it : violations_map_->DictItems()) {
+    base::Value& violation_dict = it.second;
+    base::Value* violations = violation_dict.FindKey("violations");
+    for (auto i = violations->GetList().begin();
+         i != violations->GetList().end(); i++) {
+      int64_t timestamp =
+          std::stoll(i->FindKey("timestampViolationMilliseconds")->GetString());
+      if (timestamp < cutoff_timestamp) {
+        violations->GetList().erase(i);
+        violations_count_--;
+        pending_write_ = true;
```

Comment:
Does this asynchronous write mean it's possible we delete some violations and they don't actually get deleted? It'll still eventually get deleted if that happens though, right?

Reply:
One way to think about it is the in-memory data structure is the source of truth while Watchdog is running and the on-disk json is always lagging behind and mirroring the in-memory data structure (this is to save on I/O). The on-disk json is synced to the in-memory data structure periodically and always before the on-disk json is accessed. The on-disk json is also synced before Watchdog is shutdown or triggers a crash. There is always a risk that something catastrophic occurs that causes us to lose a violation but that will be true even if we attempt write ASAP. I believe there was some discussion also about how I/O operations (especially when done very often) often poses the greatest risk to causing problems themselves. Pesky hardware.